src/Compression/Inflater.cs

```diff
@@ -0,0 +1,223 @@
+
+		/// <summary>
+		/// Initialized the Inflater with the given windowBits size
+		/// </summary>
+		internal Inflater (CompressionAlgorithm algorithm)
+		{
+			_finished = false;
+			_isDisposed = false;
+			InflateInit (algorithm);
+		}
+
+		public int AvailableOutput => (int) compression_struct.DestinationSize;
+
+		/// <summary>
+		/// Returns true if the end of the stream has been reached.
+		/// </summary>
+		public bool Finished() => _finished;
+
+		public unsafe bool Inflate (out byte b)
+		{
+			fixed (byte* bufPtr = &b)
+			{
+				int bytesRead = InflateVerified (bufPtr, 1);
+				Debug.Assert (bytesRead == 0 || bytesRead == 1);
+				return bytesRead != 0;
+			}
+		}
+
+		public unsafe int Inflate (byte[] bytes, int offset, int length)
+		{
+			// If Inflate is called on an invalid or unready inflater, return 0 to indicate no bytes have been read.
+			if (length == 0)
+				return 0;
+
+			Debug.Assert(null != bytes, "Can't pass in a null output buffer!");
+			fixed (byte* bufPtr = bytes)
+			{
+				return InflateVerified (bufPtr + offset, length);
+			}
+		}
+
+		public unsafe int Inflate (Span<byte> destination)
+		{
+			// If Inflate is called on an invalid or unready inflater, return 0 to indicate no bytes have been read.
+			if (destination.Length == 0)
+				return 0;
+
+			fixed (byte* bufPtr = &MemoryMarshal.GetReference(destination))
+			{
+				return InflateVerified (bufPtr, destination.Length);
+			}
+		}
+
+		public unsafe int InflateVerified (byte* bufPtr, int length)
+		{
+			// State is valid; attempt inflation
+			try {
+				int bytesRead;
+				var errCode = ReadInflateOutput (bufPtr, length, out bytesRead);
+				Console.WriteLine ($"ReadInflateOutput error code is {errCode }");
+				if (errCode == CompressionStatus.End) {
+					_finished = true;
+				}
+				return bytesRead;
+			} finally {
+				// Before returning, make sure to release input buffer if necessary:
+				if (0 == compression_struct.SourceSize && _inputBufferHandle.IsAllocated)
+				{
+					DeallocateInputBufferHandle ();
+				}
+			}
+		}
+
+		public bool NeedsInput() => compression_struct.SourceSize == 0;
+
+		public void SetInput(byte[] inputBuffer, int startIndex, int count)
+		{
+			Debug.Assert (NeedsInput(), "We have something left in previous input!");
+			Debug.Assert (inputBuffer != null);
+			Debug.Assert (startIndex >= 0 && count >= 0 && count + startIndex <= inputBuffer.Length);
+			Debug.Assert (!_inputBufferHandle.IsAllocated);
+
+			if (0 == count)
+				return;
+
+			lock (SyncLock)
+			{
+				_inputBufferHandle = GCHandle.Alloc(inputBuffer, GCHandleType.Pinned);
+				Console.WriteLine ("Set input");
+				compression_struct.Source = _inputBufferHandle.AddrOfPinnedObject() + startIndex;
+				compression_struct.SourceSize = count;
+				_finished = false;
+			}
+		}
+
+		private void Dispose (bool disposing)
+		{
+			if (!_isDisposed)
+			{
+				if (disposing)
+					compression_stream_destroy (ref compression_struct);
```

Comment:
same problem as before (first implementation), I don't see why `disposing` change the need of calling `compression_stream_destroy`